// @ansible/ansible-language-server/out/server/src/utils/misc.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getUnsupportedError = exports.withInterpreter = exports.insert = exports.isObject = exports.hasOwnProperty = exports.toLspRange = exports.asyncExec = exports.fileExists = void 0;
const child_process = require("child_process");
const fs_1 = require("fs");
const util_1 = require("util");
const vscode_languageserver_types_1 = require("vscode-languageserver-types");
const path = require("path");
function fileExists(filePath) {
    return __awaiter(this, void 0, void 0, function* () {
        return !!(yield fs_1.promises.stat(filePath).catch(() => false));
    });
}
exports.fileExists = fileExists;
exports.asyncExec = (0, util_1.promisify)(child_process.exec);
function toLspRange(range, textDocument) {
    const start = textDocument.positionAt(range[0]);
    const end = textDocument.positionAt(range[1]);
    return vscode_languageserver_types_1.Range.create(start, end);
}
exports.toLspRange = toLspRange;
function hasOwnProperty(obj, prop) {
    return isObject(obj) && obj.hasOwnProperty(prop);
}
exports.hasOwnProperty = hasOwnProperty;
/**
 * Checks whether `obj` is a non-null object.
 */
function isObject(obj) {
    return obj && typeof obj === "object";
}
exports.isObject = isObject;
function insert(str, index, val) {
    return `${str.substring(0, index)}${val}${str.substring(index)}`;
}
exports.insert = insert;
/**
 * Adjusts the command and environment in case the interpreter path is provided.
 */
function withInterpreter(executable, args, interpreterPath, activationScript) {
    let command = `${executable} ${args}`; // base case
    const newEnv = Object.assign({}, process.env, {
        ANSIBLE_FORCE_COLOR: "0",
        PYTHONBREAKPOINT: "0", // We want to be sure that python debugger is never
        // triggered, even if we mistakenly left a breakpoint() there while
        // debugging ansible- lint, or another tool we call.
    });
    if (activationScript) {
        command = `bash -c 'source ${activationScript} && ${executable} ${args}'`;
        return [command, undefined];
    }
    if (interpreterPath) {
        const virtualEnv = path.resolve(interpreterPath, "../..");
        const pathEntry = path.join(virtualEnv, "bin");
        if (path.isAbsolute(executable)) {
            // if both interpreter path and absolute command path are provided, we can
            // bolster the chances of success by letting the interpreter execute the
            // command
            command = `${interpreterPath} ${executable} ${args}`;
        }
        // emulating virtual environment activation script
        newEnv["VIRTUAL_ENV"] = virtualEnv;
        newEnv["PATH"] = `${pathEntry}:${process.env.PATH}`;
        delete newEnv.PYTHONHOME;
    }
    return [command, newEnv];
}
exports.withInterpreter = withInterpreter;
/**
 * Returns errors messages when LS is run on unsupported platform, or undefined
 * when all is fine.
 */
function getUnsupportedError() {
    // win32 applies to x64 arch too, is the platform name
    if (process.platform === "win32") {
        return "Ansible Language Server can only run inside WSL on Windows. Refer to vscode documentation for more details.";
    }
}
exports.getUnsupportedError = getUnsupportedError;
